### PR TITLE
DR-85 initial githooks and terraform format github action

### PR DESCRIPTION
**Pull request by timrickwood**
## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [x] I have updated the documentation accordingly
- [x] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: scripts/functions/git-functions.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    [[ ! -z "$VALID_LENGTH" ]] && echo $VALID_LENGTH
+    if [[ ! -z "$VALID_LENGTH" || ! -z "$VALID_FORMAT" ]] ; then
+      exit 1
+    fi
+}
+
+function check_git_branch_name_format {
+    BUILD_BRANCH="$1"
+    if [ $BUILD_BRANCH != 'main' ] && ! [[ $BUILD_BRANCH =~ $GIT_BRANCH_PATTERN ]]  ; then
+      echo $BUILD_BRANCH does not match the naming pattern
```
Comment (tamaragoldschmidt) [inline scripts/functions/git-functions.sh:37]: could you update the message so that it is clear it refers to a branch

Comment (timrickwood) [reply scripts/functions/git-functions.sh:37]: good shout will do

File: scripts/functions/test-git-functions.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+
+/bin/bash pre-commit-git-checks.sh task/DPTS-2211_My_valid_branch "DR-1 My message takes exactly 72 characters to describe this new commit"
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply scripts/functions/test-git-functions.sh:4]: good spot will sort out 
will probably refactor the tests

File: scripts/functions/test-terraform-functions.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+all_pass=0
```
Comment (timrickwood) [inline scripts/functions/test-terraform-functions.sh:2]: remove serves no purpose

File: scripts/git/commit-msg [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#! /bin/bash
```
Comment (timrickwood) [inline scripts/git/commit-msg:1]: change name of file to pre-commit

File: scripts/git/git-config-set-up.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# move to a git function ?
```
Comment (timrickwood) [inline scripts/git/git-config-set-up.sh:3]: change to reference pre-commit file

File: scripts/functions/test-terraform-functions.sh (hunk 3 above)
Comment (timrickwood) [reply scripts/functions/test-terraform-functions.sh:2]: Removed

File: scripts/functions/test-git-functions.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (timrickwood) [reply scripts/functions/test-git-functions.sh]: fixed in refactoring

File: scripts/githooks/terraform-pre-commit.sh [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+source ./scripts/functions/git-functions.sh
+
+TF_VERSION=1.4.6
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply scripts/githooks/terraform-pre-commit.sh:4]: sha value added

**Review by jope1 — APPROVED** [review]

**Review by gdcNHS — APPROVED** [review]

**Pull request merged**